sklearn/metrics/ranking.py

```diff
@@ -821,3 +821,251 @@ def label_ranking_loss(y_true, y_score, sample_weight=None):
     loss[np.logical_or(n_positives == 0, n_positives == n_labels)] = 0.
 
     return np.average(loss, weights=sample_weight)
+
+
+def _dcg_sample_scores(y_true, y_score, k=None, log_basis=2):
+    """Compute Discounted Cumulative Gain.
+
+    Sum the true scores ranked in the order induced by the predicted scores,
+    after applying a logarithmic discount.
+
+    This ranking metric yields a high value if true labels are ranked high by
+    ``y_score``.
+
+    Parameters
+    ----------
+    y_true : array, shape = [n_samples, n_labels], dtype = int or float
+        True targets of multilabel classification, or true scores of entities
+        to be ranked.
+
+    y_score : array, shape = [n_samples, n_labels], dtype = float
+        Target scores, can either be probability estimates, confidence values,
+        or non-thresholded measure of decisions (as returned by
+        "decision_function" on some classifiers).
+
+    k : int, optional (default=None)
+        Only consider the highest k scores in the ranking. If None, use all
+        outputs.
+
+    log_basis : float, optional (default=2)
+        Basis of the logarithm used for the discount. A low value means a
+        sharper discount (top results are more important).
+
+    Returns
+    -------
+    discounted_cumulative_gain : array, shape = [n_samples,], dtype = float
+        The DCG score for each sample.
+
+    References
+    ----------
+    `Wikipedia entry for Discounted Cumulative Gain
+        <https://en.wikipedia.org/wiki/Discounted_cumulative_gain>`_
+
+    Jarvelin, K., & Kekalainen, J. (2002).
+    Cumulated gain-based evaluation of IR techniques. ACM Transactions on
+    Information Systems (TOIS), 20(4), 422-446.
+
+    Wang, Y., Wang, L., Li, Y., He, D., Chen, W., & Liu, T. Y. (2013, May).
+    A theoretical analysis of NDCG ranking measures. In Proceedings of the 26th
+    Annual Conference on Learning Theory (COLT 2013)
+
+    See also
+    --------
+    ndcg_score :
+        The Discounted Cumulative Gain divided by the Ideal Discounted
+        Cumulative Gain (the DCG obtained for a perfect ranking), in order to
+        have a score between 0 and 1.
+
+    """
+    check_array(y_true, ensure_2d=False)
+    check_array(y_score, ensure_2d=False)
+    check_consistent_length(y_true, y_score)
+    y_type = type_of_target(y_true)
+    if y_type not in ("multilabel-indicator", "continuous-multioutput",
+                      "multiclass-multioutput"):
+        raise ValueError("{0} format is not supported".format(y_type))
+
+    ranking = np.argsort(y_score)[:, ::-1]
```

Comment:
should we be using rankdata to handle ties?

Reply:
It's true that we should handle ties.

averaging the ranks of equally scored results may not work because the summation of gains has to be cutoff @ k (we need to know how many elements of a tied group fall beyond k) . In

Computing Information Retrieval Performance Measures Efficiently in the Presence of Tied Scores. Marc Najork, Frank McSherry. ECIR, 2008

the authors average the true gains of results in a tied group before multiplying by the discount (discounts beyond k are 0)